File: 110.balanced-binary-tree.js
/*
 * @lc app=leetcode id=110 lang=javascript
 *
 * [110] Balanced Binary Tree
 */

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
  const helper = (node) => {
    if (!node) {
      return 0;
    }
    const lh = helper(node.left);
    const rh = helper(node.right);
    if (lh < 0 || rh < 0 || Math.abs(lh - rh) > 1) {
      return -1;
    }
    return 1 + Math.max(lh, rh);
  };
  return helper(root) >= 0;
};
// @lc code=end

File: 110.balanced-binary-tree.go
/*
 * @lc app=leetcode id=110 lang=golang
 *
 * [110] Balanced Binary Tree
 */

package leetcode

// @lc code=start
func isBalanced(root *TreeNode) bool {
	return helper110(root) >= 0
}

func helper110(node *TreeNode) int {
	if node == nil {
		return 0
	}
	lh, rh := helper110(node.Left), helper110(node.Right)
	if lh < 0 || rh < 0 || lh-rh > 1 || lh-rh < -1 {
		return -1
	}
	if lh > rh {
		return 1 + lh
	}
	return 1 + rh
}

// @lc code=end
